core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -794,7 +795,7 @@ class ReplicaManager(val config: KafkaConfig,
             brokerTopicStats.topicStats(topicPartition.topic).failedProduceRequestRate.mark()
             brokerTopicStats.allTopicsStats.failedProduceRequestRate.mark()
             error(s"Error processing append operation on partition $topicPartition", t)
-            (topicPartition, LogAppendResult(LogAppendInfo.unknownLogAppendInfoWithLogStartOffset(logStartOffset), Some(t)))
+            (topicPartition, LogAppendResult(LogAppendInfo.unknownLogAppendInfoWithAdditionalInfo(logStartOffset, List(), ""), Some(t)))
```

Comment:
I think we can improve this by doing the following:

1) inside LogValidator, when throwing `InvalidTimestampException` and `InvalidRecordException`, encode the relative offset inside the exception object.

2) then here we can add that relative offset as a singleton to the response.

By doing this on the client side, we can improve our error messaging such that we only tries to drop that single record but not the whole batch.